firewall/master.go

```diff
@@ -31,91 +31,170 @@ import (
 // 4. DecideOnLink
 //		is called when when the first packet of a link arrives only if communication has verdict UNDECIDED or CANTSAY
 
-// DecideOnCommunicationBeforeDNS makes a decision about a communication before the dns query is resolved and intel is gathered.
-func DecideOnCommunicationBeforeDNS(comm *network.Communication) {
+// DecideOnConnection makes a decision about a connection.
+func DecideOnConnection(conn *network.Connection, pkt packet.Packet) {
 	// update profiles and check if communication needs reevaluation
-	if comm.UpdateAndCheck() {
-		log.Infof("firewall: re-evaluating verdict on %s", comm)
-		comm.ResetVerdict()
+	if conn.UpdateAndCheck() {
+		log.Infof("filter: re-evaluating verdict on %s", conn)
+		conn.Verdict = network.VerdictUndecided
 	}
 
-	// check if need to run
-	if comm.GetVerdict() != network.VerdictUndecided {
+	// grant self
+	if conn.Process().Pid == os.Getpid() {
+		log.Infof("filter: granting own connection %s", conn)
+		conn.Verdict = network.VerdictAccept
 		return
 	}
 
-	// grant self
-	if comm.Process().Pid == os.Getpid() {
-		log.Infof("firewall: granting own communication %s", comm)
-		comm.Accept("")
-		return
+	// check if process is communicating with itself
+	if pkt != nil {
+		if conn.Process().Pid >= 0 && pkt.Info().Src.Equal(pkt.Info().Dst) {
+			// get PID
+			otherPid, _, err := process.GetPidByEndpoints(
+				pkt.Info().RemoteIP(),
+				pkt.Info().RemotePort(),
+				pkt.Info().LocalIP(),
+				pkt.Info().LocalPort(),
+				pkt.Info().Protocol,
+			)
+			if err == nil {
+
+				// get primary process
+				otherProcess, err := process.GetOrFindPrimaryProcess(pkt.Ctx(), otherPid)
+				if err == nil {
+
+					if otherProcess.Pid == conn.Process().Pid {
+						conn.Accept("connection to self")
+						return
+					}
+				}
+			}
+		}
```

Comment:
`process.GetPidByEndpoints` and `process.GetOrFindPrimaryProcess` can fail for various reasons and we completely hide those errors here. I guess there are only a few (if not only one) that we can safely hide. May add some logging for the others.

Reply:
Add logging.